src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -0,0 +1,1355 @@
+    searchHandle_ = handle;
+}
+
+//
+// EspIdfWiFi::mdns_start()
+//
+StateFlowBase::Action EspIdfWiFi::mdns_start()
+{
+    HASSERT(mdnsClientCache_.size() > 0);
+    if (mdnsAdvInhibitSta_)
+    {
+        mdns_adv_inhibit();
+        mdns_restore_sta();
+        mdnsAdvInhibitStaActive_ = true;
+    }
+    return call_immediately(STATE(mdns_query));
+}
+
+//
+// EspIdfWiFi::mdns_query()
+//
+StateFlowBase::Action EspIdfWiFi::mdns_query()
+{
+    OSMutexLock lock(this);
+    mdnsClientTrigRefresh_ = false;
+    for (auto it = mdnsClientCache_.begin(); it != mdnsClientCache_.end(); ++it)
+    {
+        std::string name;
+        std::string proto;
+        mdns_split((*it).service_.c_str(), &name, &proto);
+        (*it).reset((void*)mdns_query_async_new(
+            nullptr, name.c_str(), proto.c_str(), MDNS_TYPE_PTR,
+            MDNS_QUERY_ASYNC_NEW_TIMEOUT_MSEC, 3, nullptr));
+        LOG(VERBOSE, "wifi: mDNS new search query: %s.%s",
+            name.c_str(), proto.c_str());
+    }
+
+    return sleep_and_call(&timer_,
+        MDNS_QUERY_CHECK_TIMEOUT_NSEC, STATE(mdns_check));
+}
+
+//
+// EspIdfWiFi::mdns_check()
+//
+StateFlowBase::Action EspIdfWiFi::mdns_check()
+{
+    bool search_still_active = false;
+    OSMutexLock lock(this);
+    // Loop through all of the service clients.
+    for (auto it = mdnsClientCache_.begin(); it != mdnsClientCache_.end(); ++it)
+    {
+        if ((*it).get() == nullptr)
+        {
+            // No active search for this cache entry.
+            LOG(VERBOSE, "wifi: mDNS no active search for this cache entry.");
+            continue;
+        }
+        mdns_result_t *mdns_result = nullptr;
+        bool result = mdns_query_async_get_results(
+            (mdns_search_once_t*)(*it).get(), 0, &mdns_result, nullptr);
+        if (!result)
+        {
+            // Results not ready yet.
+            LOG(VERBOSE, "wifi: mDNS client results not ready yet.");
+            search_still_active = true;
+            continue;
+        }
+        if (mdns_result == nullptr)
+        {
+            // No results for this service.
+            LOG(VERBOSE, "wifi: No mDNS client results for this service.");
+            (*it).reset();
+            continue;
+        }
+
+        // Loop through all the results.
+        for (mdns_result_t *cur_result = mdns_result;
+            cur_result; cur_result = cur_result->next)
+        {
+            MDNSCacheAddr cache_addr;
+            cache_addr.port_ = htons(cur_result->port);
+
+            for (mdns_ip_addr_t *addr = cur_result->addr; addr;
+                addr = addr->next)
+            {
+                // For some reason, the mdns_result_t::ip_protocol cannot
+                // be trusted. The address results list can contain both
+                // IPv4 and IPv6 addresses.
+                if (addr->addr.type == ESP_IPADDR_TYPE_V4)
+                {
+                    cache_addr.family_ = AF_INET;
+                }
+#if defined(ESP_IDF_WIFI_IPV6)
```

Comment:
`CONFIG_LWIP_IPV6` might be a good replacement here, with possible fallback to force IPv4